ICE: `can't type-check body of DefId`

<!--
[31mICE[0m: Rustc ./ADE735FE94A56A88EF58720151A6E182A564F05FC77122B9F99A8E35987153B1.rs '' 'error: internal compiler error: compiler/rustc_hir_typeck/src/lib.rs:127:9: can't type-check body of DefId(0:8 ~ ADE735FE94A56A88EF58720151A6E182A564F05FC77122B9F99A8E35987153B1[31c4]::{impl#0}::NUM)', 'error: internal compiler error: compiler/rustc_hir_typeck/src/lib.rs:127:9: can't type-check body of DefId(0:8 ~ ADE735FE94A56A88EF58720151A6E182A564F05FC77122B9F99A8E35987153B1[31c4]::{impl#0}::NUM)'
File: /tmp/im/ADE735FE94A56A88EF58720151A6E182A564F05FC77122B9F99A8E35987153B1.rs
-->

auto-reduced (treereduce-rust):
````rust
trait BadTr {
    
}

impl BadTr for GoodS {
    #[type_const]
    const NUM:  = 84;
}

````

original:
````rust
#![expect(incomplete_features)]
#![feature(associated_const_equality, min_generic_const_args)]

trait BadTr {
    const NUM: usize;
}

struct GoodS;

impl BadTr for GoodS {
    #[type_const]
    const NUM:  = 84;
}

fn accept_bad_tr<const N: usize, T: BadTr<NUM = { N }>>(_x: &T) {}
//~^ ERROR use of trait associated const without `#[type_const]`

fn main() {
    accept_bad_tr::<GoodS = { N }>(&GoodS);
}

````

Version information
````
rustc 1.93.0-nightly (acda5e9f9 2025-11-09)
binary: rustc
commit-hash: acda5e9f9aadac705b851935b14aa3d34a67d428
commit-date: 2025-11-09
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.3

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/acda5e9f9aadac705b851935b14aa3d34a67d428/compiler/rustc_hir_typeck/src/lib.rs#L121-L133

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0438]: const `NUM` is not a member of trait `BadTr`
  --> /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:12:5
   |
12 |     const NUM:  = 84;
   |     ^^^^^^^^^^^^^^^^^ not a member of trait `BadTr`

error[E0412]: cannot find type `GoodS` in this scope
  --> /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:10:16
   |
10 | impl BadTr for GoodS {
   |                ^^^^^ not found in this scope

error[E0658]: the `#[type_const]` attribute is an experimental feature
  --> /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:11:5
   |
11 |     #[type_const]
   |     ^^^^^^^^^^^^^
   |
   = note: see issue #132980 <https://github.com/rust-lang/rust/issues/132980> for more information
   = help: add `#![feature(min_generic_const_args)]` to the crate attributes to enable
   = note: this compiler was built on 2025-11-09; consider upgrading it if it is out of date

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:13:2
   |
13 | }
   |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs`

error: internal compiler error: compiler/rustc_hir_typeck/src/lib.rs:127:9: can't type-check body of DefId(0:5 ~ mvce[3608]::{impl#0}::NUM)
  --> /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:12:5
   |
12 |     const NUM:  = 84;
   |     ^^^^^^^^^^


thread 'rustc' (3860575) panicked at compiler/rustc_hir_typeck/src/lib.rs:127:9:
Box<dyn Any>
stack backtrace:
   0:     0x7f2b3026c9c3 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h0dc93c2810c40bd0
   1:     0x7f2b30a01afc - core::fmt::write::hacdaf2bbd154e9b5
   2:     0x7f2b30221533 - std::io::Write::write_fmt::h75e948926e9dfbe3
   3:     0x7f2b30232b42 - std::sys::backtrace::BacktraceLock::print::h251013828edeee18
   4:     0x7f2b30238b99 - std::panicking::default_hook::{{closure}}::hf6c9462acd913055
   5:     0x7f2b302386c3 - std::panicking::default_hook::h4c9fd5029d2f8b73
   6:     0x7f2b2f22a721 - std[b4c06921f787550c]::panicking::update_hook::<alloc[7ed4a9fe7ec582b8]::boxed::Box<rustc_driver_impl[87c4f5c8377f9862]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7f2b30238fbf - std::panicking::panic_with_hook::h02a757165762df79
   8:     0x7f2b2f266cb1 - std[b4c06921f787550c]::panicking::begin_panic::<rustc_errors[373a255116eccc4c]::ExplicitBug>::{closure#0}
   9:     0x7f2b2f2593d6 - std[b4c06921f787550c]::sys::backtrace::__rust_end_short_backtrace::<std[b4c06921f787550c]::panicking::begin_panic<rustc_errors[373a255116eccc4c]::ExplicitBug>::{closure#0}, !>
  10:     0x7f2b2f256a94 - std[b4c06921f787550c]::panicking::begin_panic::<rustc_errors[373a255116eccc4c]::ExplicitBug>
  11:     0x7f2b2f285321 - <rustc_errors[373a255116eccc4c]::diagnostic::BugAbort as rustc_errors[373a255116eccc4c]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:     0x7f2b2f7ea57c - <rustc_errors[373a255116eccc4c]::DiagCtxtHandle>::span_bug::<rustc_span[39972e28049286de]::span_encoding::Span, alloc[7ed4a9fe7ec582b8]::string::String>
  13:     0x7f2b2f80da87 - rustc_middle[1e1513fa521466b3]::util::bug::opt_span_bug_fmt::<rustc_span[39972e28049286de]::span_encoding::Span>::{closure#0}
  14:     0x7f2b2f80dc2a - rustc_middle[1e1513fa521466b3]::ty::context::tls::with_opt::<rustc_middle[1e1513fa521466b3]::util::bug::opt_span_bug_fmt<rustc_span[39972e28049286de]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  15:     0x7f2b2f7fe63b - rustc_middle[1e1513fa521466b3]::ty::context::tls::with_context_opt::<rustc_middle[1e1513fa521466b3]::ty::context::tls::with_opt<rustc_middle[1e1513fa521466b3]::util::bug::opt_span_bug_fmt<rustc_span[39972e28049286de]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  16:     0x7f2b2db06757 - rustc_middle[1e1513fa521466b3]::util::bug::span_bug_fmt::<rustc_span[39972e28049286de]::span_encoding::Span>
  17:     0x7f2b31bd6e1a - rustc_hir_typeck[b7947f0b53f42892]::typeck_with_inspect::{closure#0}
  18:     0x7f2b31bcf00e - rustc_query_impl[d4db0883806fa552]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d4db0883806fa552]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1e1513fa521466b3]::query::erase::Erased<[u8; 8usize]>>
  19:     0x7f2b30c53677 - rustc_query_system[d03d33f2f2d15895]::query::plumbing::try_execute_query::<rustc_query_impl[d4db0883806fa552]::DynamicConfig<rustc_data_structures[aff1d8a073485c2a]::vec_cache::VecCache<rustc_span[39972e28049286de]::def_id::LocalDefId, rustc_middle[1e1513fa521466b3]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[d03d33f2f2d15895]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[d4db0883806fa552]::plumbing::QueryCtxt, false>
  20:     0x7f2b30c53009 - rustc_query_impl[d4db0883806fa552]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  21:     0x7f2b2f3f11c9 - rustc_hir_analysis[3bbcf63e422e422d]::collect::type_of::infer_placeholder_type
  22:     0x7f2b313f127e - rustc_hir_analysis[3bbcf63e422e422d]::collect::type_of::type_of
  23:     0x7f2b30a1d926 - rustc_query_impl[d4db0883806fa552]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d4db0883806fa552]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1e1513fa521466b3]::query::erase::Erased<[u8; 8usize]>>
  24:     0x7f2b30a1c5c6 - rustc_query_system[d03d33f2f2d15895]::query::plumbing::try_execute_query::<rustc_query_impl[d4db0883806fa552]::DynamicConfig<rustc_query_system[d03d33f2f2d15895]::query::caches::DefIdCache<rustc_middle[1e1513fa521466b3]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[d4db0883806fa552]::plumbing::QueryCtxt, false>
  25:     0x7f2b30a1c16f - rustc_query_impl[d4db0883806fa552]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  26:     0x7f2b3120ddd9 - rustc_hir_analysis[3bbcf63e422e422d]::check::check::check_item_type
  27:     0x7f2b3120c41c - rustc_hir_analysis[3bbcf63e422e422d]::check::wfcheck::check_well_formed
  28:     0x7f2b3120c3f5 - rustc_query_impl[d4db0883806fa552]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d4db0883806fa552]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1e1513fa521466b3]::query::erase::Erased<[u8; 1usize]>>
  29:     0x7f2b3120bc6c - rustc_query_system[d03d33f2f2d15895]::query::plumbing::try_execute_query::<rustc_query_impl[d4db0883806fa552]::DynamicConfig<rustc_data_structures[aff1d8a073485c2a]::vec_cache::VecCache<rustc_span[39972e28049286de]::def_id::LocalDefId, rustc_middle[1e1513fa521466b3]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[d03d33f2f2d15895]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[d4db0883806fa552]::plumbing::QueryCtxt, false>
  30:     0x7f2b3120b782 - rustc_query_impl[d4db0883806fa552]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  31:     0x7f2b31208d08 - rustc_hir_analysis[3bbcf63e422e422d]::check::wfcheck::check_type_wf
  32:     0x7f2b31208bf9 - rustc_query_impl[d4db0883806fa552]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d4db0883806fa552]::query_impl::check_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1e1513fa521466b3]::query::erase::Erased<[u8; 1usize]>>
  33:     0x7f2b31a47dd4 - rustc_query_system[d03d33f2f2d15895]::query::plumbing::try_execute_query::<rustc_query_impl[d4db0883806fa552]::DynamicConfig<rustc_query_system[d03d33f2f2d15895]::query::caches::SingleCache<rustc_middle[1e1513fa521466b3]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[d4db0883806fa552]::plumbing::QueryCtxt, false>
  34:     0x7f2b31a47bb6 - rustc_query_impl[d4db0883806fa552]::query_impl::check_type_wf::get_query_non_incr::__rust_end_short_backtrace
  35:     0x7f2b3152ea3a - rustc_hir_analysis[3bbcf63e422e422d]::check_crate
  36:     0x7f2b30c497b0 - rustc_interface[df9ddd9c9467ef8b]::passes::analysis
  37:     0x7f2b30c49465 - rustc_query_impl[d4db0883806fa552]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d4db0883806fa552]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[1e1513fa521466b3]::query::erase::Erased<[u8; 0usize]>>
  38:     0x7f2b31a4e814 - rustc_query_system[d03d33f2f2d15895]::query::plumbing::try_execute_query::<rustc_query_impl[d4db0883806fa552]::DynamicConfig<rustc_query_system[d03d33f2f2d15895]::query::caches::SingleCache<rustc_middle[1e1513fa521466b3]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[d4db0883806fa552]::plumbing::QueryCtxt, false>
  39:     0x7f2b31a4e44a - rustc_query_impl[d4db0883806fa552]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  40:     0x7f2b31d09f83 - <rustc_interface[df9ddd9c9467ef8b]::passes::create_and_enter_global_ctxt<core[f1f7c8de62d6692e]::option::Option<rustc_interface[df9ddd9c9467ef8b]::queries::Linker>, rustc_driver_impl[87c4f5c8377f9862]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[f1f7c8de62d6692e]::ops::function::FnOnce<(&rustc_session[5dff62ba5cbc554d]::session::Session, rustc_middle[1e1513fa521466b3]::ty::context::CurrentGcx, alloc[7ed4a9fe7ec582b8]::sync::Arc<rustc_data_structures[aff1d8a073485c2a]::jobserver::Proxy>, &std[b4c06921f787550c]::sync::once_lock::OnceLock<rustc_middle[1e1513fa521466b3]::ty::context::GlobalCtxt>, &rustc_data_structures[aff1d8a073485c2a]::sync::worker_local::WorkerLocal<rustc_middle[1e1513fa521466b3]::arena::Arena>, &rustc_data_structures[aff1d8a073485c2a]::sync::worker_local::WorkerLocal<rustc_hir[2565f01ea710f611]::Arena>, rustc_driver_impl[87c4f5c8377f9862]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  41:     0x7f2b31b8ec69 - rustc_interface[df9ddd9c9467ef8b]::interface::run_compiler::<(), rustc_driver_impl[87c4f5c8377f9862]::run_compiler::{closure#0}>::{closure#1}
  42:     0x7f2b3188c98f - std[b4c06921f787550c]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[df9ddd9c9467ef8b]::util::run_in_thread_with_globals<rustc_interface[df9ddd9c9467ef8b]::util::run_in_thread_pool_with_globals<rustc_interface[df9ddd9c9467ef8b]::interface::run_compiler<(), rustc_driver_impl[87c4f5c8377f9862]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  43:     0x7f2b3188c670 - <<std[b4c06921f787550c]::thread::Builder>::spawn_unchecked_<rustc_interface[df9ddd9c9467ef8b]::util::run_in_thread_with_globals<rustc_interface[df9ddd9c9467ef8b]::util::run_in_thread_pool_with_globals<rustc_interface[df9ddd9c9467ef8b]::interface::run_compiler<(), rustc_driver_impl[87c4f5c8377f9862]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[f1f7c8de62d6692e]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  44:     0x7f2b3188afef - std::sys::thread::unix::Thread::new::thread_start::h6ee5f4a2e2de181d
  45:     0x7f2b2b2969cb - <unknown>
  46:     0x7f2b2b31aa0c - <unknown>
  47:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.93.0-nightly (acda5e9f9 2025-11-09) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `<impl at /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:10:1: 10:21>::NUM`
#1 [type_of] computing type of `<impl at /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:10:1: 10:21>::NUM`
#2 [check_well_formed] checking that `<impl at /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:10:1: 10:21>::NUM` is well-formed
#3 [check_type_wf] checking that types are well-formed
#4 [analysis] running analysis passes on crate `mvce`
end of query stack
error: missing type for `const` item
  --> /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:12:15
   |
12 |     const NUM:  = 84;
   |               ^
   |
help: provide a type for the item
   |
12 |     const NUM: <type>  = 84;
   |                ++++++

error: aborting due to 6 previous errors

Some errors have detailed explanations: E0412, E0438, E0601, E0658.
For more information about an error, try `rustc --explain E0412`.

```

</p>
</details>

<!--
query stack:
#0 [typeck] type-checking `<impl at /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:10:1: 10:21>::NUM`
#1 [type_of] computing type of `<impl at /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:10:1: 10:21>::NUM`
#2 [check_well_formed] checking that `<impl at /tmp/icemaker_global_tempdir.pcYJpEqsAGEB/rustc_testrunner_tmpdir_reporting.d7A4ynwfWceP/mvce.rs:10:1: 10:21>::NUM` is well-formed
#3 [check_type_wf] checking that types are well-formed
#4 [analysis] running analysis passes on crate `mvce`

-->
@rustbot label +F-associated_const_equality +F-min_generic_const_args

--- a/compiler/rustc_const_eval/src/check_consts/qualifs.rs
+++ b/compiler/rustc_const_eval/src/check_consts/qualifs.rs
@@ -6,9 +6,7 @@
 // having basically only two use-cases that act in different ways.
 
 use rustc_errors::ErrorGuaranteed;
-use rustc_hir::attrs::AttributeKind;
-use rustc_hir::def::DefKind;
-use rustc_hir::{LangItem, find_attr};
+use rustc_hir::LangItem;
 use rustc_infer::infer::TyCtxtInferExt;
 use rustc_middle::mir::*;
 use rustc_middle::ty::{self, AdtDef, Ty};
@@ -366,14 +364,10 @@ where
         // check performed after the promotion. Verify that with an assertion.
         assert!(promoted.is_none() || Q::ALLOW_PROMOTED);
 
-        // Avoid looking at attrs of anon consts as that will ICE
-        let is_type_const_item =
-            matches!(cx.tcx.def_kind(def), DefKind::Const | DefKind::AssocConst)
-                && find_attr!(cx.tcx.get_all_attrs(def), AttributeKind::TypeConst(_));
-
         // Don't peak inside trait associated constants, also `#[type_const] const` items
         // don't have bodies so there's nothing to look at
-        if promoted.is_none() && cx.tcx.trait_of_assoc(def).is_none() && !is_type_const_item {
+        if promoted.is_none() && cx.tcx.trait_of_assoc(def).is_none() && !cx.tcx.is_type_const(def)
+        {
             let qualifs = cx.tcx.at(constant.span).mir_const_qualif(def);
 
             if !Q::in_qualifs(&qualifs) {

--- a/compiler/rustc_hir_analysis/src/collect/type_of.rs
+++ b/compiler/rustc_hir_analysis/src/collect/type_of.rs
@@ -420,7 +420,22 @@ fn infer_placeholder_type<'tcx>(
     kind: &'static str,
 ) -> Ty<'tcx> {
     let tcx = cx.tcx();
-    let ty = tcx.typeck(def_id).node_type(hir_id);
+    // If the type is omitted on a #[type_const] we can't run
+    // type check on since that requires the const have a body
+    // which type_consts don't.
+    let ty = if tcx.is_type_const(def_id.to_def_id()) {
+        if let Some(trait_item_def_id) = tcx.trait_item_of(def_id.to_def_id()) {
+            tcx.type_of(trait_item_def_id).instantiate_identity()
+        } else {
+            Ty::new_error_with_message(
+                tcx,
+                ty_span,
+                "constant with #[type_const] requires an explicit type",
+            )
+        }
+    } else {
+        tcx.typeck(def_id).node_type(hir_id)
+    };
 
     // If this came from a free `const` or `static mut?` item,
     // then the user may have written e.g. `const A = 42;`.

--- a/compiler/rustc_middle/src/ty/context.rs
+++ b/compiler/rustc_middle/src/ty/context.rs
@@ -1891,6 +1891,12 @@ impl<'tcx> TyCtxt<'tcx> {
         self.is_lang_item(self.parent(def_id), LangItem::AsyncDropInPlace)
     }
 
+    /// Check if the given `def_id` is a const with the `#[type_const]` attribute.
+    pub fn is_type_const(self, def_id: DefId) -> bool {
+        matches!(self.def_kind(def_id), DefKind::Const | DefKind::AssocConst)
+            && find_attr!(self.get_all_attrs(def_id), AttributeKind::TypeConst(_))
+    }
+
     /// Returns the movability of the coroutine of `def_id`, or panics
     /// if given a `def_id` that is not a coroutine.
     pub fn coroutine_movability(self, def_id: DefId) -> hir::Movability {

--- a/compiler/rustc_mir_build/src/builder/expr/as_constant.rs
+++ b/compiler/rustc_mir_build/src/builder/expr/as_constant.rs
@@ -1,7 +1,7 @@
 //! See docs in build/expr/mod.rs
 
 use rustc_abi::Size;
-use rustc_ast as ast;
+use rustc_ast::{self as ast};
 use rustc_hir::LangItem;
 use rustc_middle::mir::interpret::{CTFE_ALLOC_SALT, LitToConstInput, Scalar};
 use rustc_middle::mir::*;
@@ -47,6 +47,7 @@ pub(crate) fn as_constant_inner<'tcx>(
     tcx: TyCtxt<'tcx>,
 ) -> ConstOperand<'tcx> {
     let Expr { ty, temp_scope_id: _, span, ref kind } = *expr;
+
     match *kind {
         ExprKind::Literal { lit, neg } => {
             let const_ = lit_to_mir_constant(tcx, LitToConstInput { lit: lit.node, ty, neg });
@@ -69,6 +70,13 @@ pub(crate) fn as_constant_inner<'tcx>(
         }
         ExprKind::NamedConst { def_id, args, ref user_ty } => {
             let user_ty = user_ty.as_ref().and_then(push_cuta);
+            if tcx.is_type_const(def_id) {
+                let uneval = ty::UnevaluatedConst::new(def_id, args);
+                let ct = ty::Const::new_unevaluated(tcx, uneval);
+
+                let const_ = Const::Ty(ty, ct);
+                return ConstOperand { span, user_ty, const_ };
+            }
 
             let uneval = mir::UnevaluatedConst::new(def_id, args);
             let const_ = Const::Unevaluated(uneval, ty);

--- a/tests/ui/const-generics/mgca/type_const-array-return.rs
+++ b/tests/ui/const-generics/mgca/type_const-array-return.rs
@@ -1,0 +1,26 @@
+//@ check-pass
+// This test should compile without an ICE.
+#![expect(incomplete_features)]
+#![feature(min_generic_const_args)]
+
+pub struct A;
+
+pub trait Array {
+    #[type_const]
+    const LEN: usize;
+    fn arr() -> [u8; Self::LEN];
+}
+
+impl Array for A {
+    #[type_const]
+    const LEN: usize = 4;
+
+    #[allow(unused_braces)]
+    fn arr() -> [u8; const { Self::LEN }] {
+        return [0u8; const { Self::LEN }];
+    }
+}
+
+fn main() {
+    let _ = A::arr();
+}

--- a/tests/ui/const-generics/mgca/type_const-inherent-const-omitted-type.rs
+++ b/tests/ui/const-generics/mgca/type_const-inherent-const-omitted-type.rs
@@ -1,0 +1,12 @@
+#![expect(incomplete_features)]
+#![feature(min_generic_const_args)]
+
+struct A;
+
+impl A {
+    #[type_const]
+    const B = 4;
+    //~^ ERROR: missing type for `const` item
+}
+
+fn main() {}

--- a/tests/ui/const-generics/mgca/type_const-inherent-const-omitted-type.stderr
+++ b/tests/ui/const-generics/mgca/type_const-inherent-const-omitted-type.stderr
@@ -1,0 +1,13 @@
+error: missing type for `const` item
+  --> $DIR/type_const-inherent-const-omitted-type.rs:8:12
+   |
+LL |     const B = 4;
+   |            ^
+   |
+help: provide a type for the item
+   |
+LL |     const B: <type> = 4;
+   |            ++++++++
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/const-generics/mgca/type_const-only-in-impl-omitted-type.rs
+++ b/tests/ui/const-generics/mgca/type_const-only-in-impl-omitted-type.rs
@@ -1,0 +1,22 @@
+#![expect(incomplete_features)]
+#![feature(min_generic_const_args)]
+
+trait BadTr {
+    const NUM: usize;
+}
+
+struct GoodS;
+
+impl BadTr for GoodS {
+    #[type_const]
+    const NUM: = 84;
+    //~^ ERROR: missing type for `const` item
+
+}
+
+fn accept_bad_tr<const N: usize, T: BadTr<NUM = { N }>>(_x: &T) {}
+//~^ ERROR use of trait associated const without `#[type_const]`
+
+fn main() {
+    accept_bad_tr::<84, _>(&GoodS);
+}

--- a/tests/ui/const-generics/mgca/type_const-only-in-impl-omitted-type.stderr
+++ b/tests/ui/const-generics/mgca/type_const-only-in-impl-omitted-type.stderr
@@ -1,0 +1,16 @@
+error: missing type for `const` item
+  --> $DIR/type_const-only-in-impl-omitted-type.rs:12:15
+   |
+LL |     const NUM: = 84;
+   |               ^ help: provide a type for the associated constant: `usize`
+
+error: use of trait associated const without `#[type_const]`
+  --> $DIR/type_const-only-in-impl-omitted-type.rs:17:43
+   |
+LL | fn accept_bad_tr<const N: usize, T: BadTr<NUM = { N }>>(_x: &T) {}
+   |                                           ^^^^^^^^^^^
+   |
+   = note: the declaration in the trait must be marked with `#[type_const]`
+
+error: aborting due to 2 previous errors
+

--- a/tests/ui/const-generics/mgca/type_const-recursive.rs
+++ b/tests/ui/const-generics/mgca/type_const-recursive.rs
@@ -1,0 +1,8 @@
+#![expect(incomplete_features)]
+#![feature(min_generic_const_args)]
+
+#[type_const]
+const A: u8 = A;
+//~^ ERROR: overflow normalizing the unevaluated constant `A` [E0275]
+
+fn main() {}

--- a/tests/ui/const-generics/mgca/type_const-recursive.stderr
+++ b/tests/ui/const-generics/mgca/type_const-recursive.stderr
@@ -1,0 +1,11 @@
+error[E0275]: overflow normalizing the unevaluated constant `A`
+  --> $DIR/type_const-recursive.rs:5:1
+   |
+LL | const A: u8 = A;
+   | ^^^^^^^^^^^
+   |
+   = note: in case this is a recursive type alias, consider using a struct, enum, or union instead
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0275`.

--- a/tests/ui/const-generics/mgca/type_const-use.rs
+++ b/tests/ui/const-generics/mgca/type_const-use.rs
@@ -1,0 +1,13 @@
+//@ check-pass
+// This test should compile without an ICE.
+#![expect(incomplete_features)]
+#![feature(min_generic_const_args)]
+
+#[type_const]
+const CONST: usize = 1;
+
+fn uses_const() {
+    CONST;
+}
+
+fn main() {}

